### ISSUE TITLE
Invalid key/Decryption error between 2 Linux clients

### ISSUE THREAD

**Issue by haelix888** [meganz/MEGAcmd#557]
2 Linux clients, A uploads B downloads, both version megacmd 1.4.0

Client A uploads, client B does this:

```
[API:debug: 23:00:05] Contiguous progress: 5425190 (0)
[API:warn: 23:00:05] Event 99431: MAC verification failed
[API:debug: 23:00:05] Transfer failed with error -14
[API:warn: 23:00:05] Event 99433: Undecryptable file
[API:debug: 23:00:05] Removing transfer
[API:warn: 23:00:05] Transfer (DOWNLOAD) finished with error: Invalid key/Decryption error File: asi se canta.mp3
[API:debug: 23:00:05] Removing cached transfer

```

Client B then removes the file and the file is gone from the cloud and from client A!
Can someone please diagnose, this is critical. I can help reproduce.

**Comment by haelix888** [comment]
One relevant observation, I believe.
If the file is uploaded with a **_different timestamp_**, client B no longer rejects it. There must be a bug in the checksum calculation (which includes file attributes)??

**Comment by haelix888** [comment]
I found minimal steps to reproduce. Please do the following.
-------
1. Unpack the file 'yo no se porque te quiero.mp3' (this is Spanish for 'I don't know why I love you') to a location untracked by MEGA; do NOT rename it.
[yo.zip](https://github.com/meganz/MEGAcmd/files/6950679/yo.zip) (md5sum is 1d6076a297498995492fa82d6f2a0eba)

2. Set the modification time to this magical value:
`touch -m -t 201212131218.50 ~/yo\ no\ se\ porque\ te\ quiero.mp3`
Test with
`stat ~/yo\ no\ se\ porque\ te\ quiero.mp3`

3. Copy to a folder being synced by mega-cmd, in attributes-preserving fashion (-a flag):
`cp -a ~/yo\ no\ se\ porque\ te\ quiero.mp3 mySyncedFolder/`

4. A different computer running mega-cmd will fail to decrypt the file, and remove it from the cloud, outputting the below messages
```
[API:debug: 23:00:05] Contiguous progress: 5425190 (0)
[API:warn: 23:00:05] Event 99431: MAC verification failed
[API:debug: 23:00:05] Transfer failed with error -14
[API:warn: 23:00:05] Event 99433: Undecryptable file
[API:debug: 23:00:05] Removing transfer
[API:warn: 23:00:05] Transfer (DOWNLOAD) finished with error: Invalid key/Decryption error File: yo no se porque te quiero.mp3
[API:debug: 23:00:05] Removing cached transfer
```
5. To make it work, alter modification timestamp by just ONE SECOND, as below. Copy as in Step 3 and the second computer will accept it.
`touch -m -t 201212131218.51 ~/yo\ no\ se\ porque\ te\ quiero.mp3`

6. (_**Alternative way to make it work**_) Leave the timestamp alone, but change the name to 'yo se porque te quiero.mp3' (I removed the negation). Copy as in Step 3 and the second computer will accept it.


### Anyone keen to hunt for this bad boy?

**Comment by mattw-mega** [comment]
Do you already have a copy of that exact file with that exact timestamp in your cloud account?   In that case the "upload" would duplicate the cloud node instead of uploading the file afresh (which saves bandwidth).  That would be quite a different scenario than the timestamp somehow causing a corrupt upload, and would explain this case (though, it is unfortunate that the file already in the cloud has that problem).  thanks

**Comment by haelix888** [comment]
@mattw-mega no, I do not. I'm aware of the behavior you describe when uploading copies of already-existing files, and my tests were done with fresh files.

I can also confirm that if a file is uploaded, synced on a 2nd computer, and then deleted (goes to debris), re-uploading it will NOT trigger the optimisation you describe. I.e., the file is uploaded anew and also on the 2nd computer it's downloaded anew, rather than "resurrected" from local debris folder. In other words, I was able to reproduce the bug as many times as I wanted by just removing the file and re-copying it into the mega-synced location.

**Comment by dwmega** [comment]
Hey @haelix888,

I've followed your instructions and unfortunately, have been unable to reproduce the issue.

What distribution and filesystem are you using?

Does the issue occur if you use two different instances of mega-cmd on the same machine?
Say, two clients running under distinct Linux users?

**Comment by haelix888** [comment]
Thanks @dwmega 

And now for the really bad news - I cannot reproduce it myself today, after a system upgrade.
My mega-cmd run on different computers, but inside the same docker image, which I did NOT upgrade.
I tried also after clearing the Trash bin in the cloud (using a browser) as well as locally on both machines.

I'm positive that the bug was occurring deterministically when changing *just* the timestamp as described, and it was symmetric (it happened when uploading on either computer and downloading on the other).
Another thing I had checked - if useful - was that, when uploading from machine A, an Android client (MegaSync) happily downloaded the offending file, but as soon as machine B came online it was reporting the Decryption error. To me this further suggests a mega-cmd issue.

**Comment by haelix888** [comment]
FYI I tried downgrading all packages including the kernel (both machines) and still cannot reproduce.
